api.yaml

```diff
@@ -0,0 +1,348 @@
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Finding'
+        "500":
+          description: "Internal server error"
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Error'
+      x-transportd:
+        backend: app
+        enabled:
+          - "accesslog"
+          - "requestvalidation"
+          - "responsevalidation"
+          - "lambda"
+        lambda:
+          arn: "results"
+          async: false
+          request: '{"id": "#! .Request.URL.resultId !#"}'
+          success: >
+            #! if eq .Response.Body.status "IN_PROGRESS" !#
+            {"status":204}
+            #! else if eq .Response.Body.status "UNKNOWN" !#
+            {"status":404}
+            #! else if eq .Response.Body.status "READY" !#
+            {"status":200,"body": {"findings": #! json .Response.Body.findings #!}
+            #! end !#
+          error: '{"status": 500, "bodyPassthrough": true}'
+  /schedule:
+    post:
+      description: Schedule a host to be scanned
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/ScanRequest'
+      responses:
+        "202":
+          description: "Success"
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ScanScheduleResponse'
+        "400":
+          description: "Invalid input"
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Error'
+      x-transportd:
+        backend: app
+        enabled:
+          - "accesslog"
+          - "requestvalidation"
+          - "responsevalidation"
+          - "lambda"
+        lambda:
+          arn: "schedule"
+          async: false
+          request: '#! json .Request.Body !#'
+          success: '{"status": 200, "bodyPassthrough": true}'
+          error: '{"status": 400, "bodyPassthrough": true}'
+components:
+  schemas:
+    Finding:
+      type: object
+      properties:
+        timestamp:
+          type: string
+          description: The time at which the scan was performed.
+        ip:
+          type: string
+          description: The IP address scanned
+        hostnames:
+          type: array
+          items:
+            type: string
+          description: Any known hostnames associated with the IP
+        vulnerabilities:
+          type: array
+          items:
+            $ref: '#/components/schemas/Vulnerability'
+    Vulnerability:
+      type: object
+      properties:
+        key:
+          type: string
+          description: Unique identity of the vulnerability as reported by nmap.
+        title:
+          type: string
+          description: Name of the vulnerability.
+        state:
+          type: string
+          description: >
```

Comment:
Question: why not use an enum here?

Reply:
Because enums can't be modified in any way, including additions, without breaking the API contract. I was hoping to sidestep that for a bit while I thought about how to handle this. There are some proposals for OpenAPI/JSONSchema around "extendable enums" but none of them are official yet.